In [48]:
import pandas as pd
import numpy as np

In [49]:
df = pd.read_csv('C:\\Users\\stolb\\Desktop\\Материалы к лекциям\\Статистика в Python\\Лекция 1\\horse_data.txt', sep = ',', na_values = '?', header = None)

df.head()

0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0  2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1  1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2  2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3  1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4  2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   

    20   21   22  23     24  25  26  27  
0  NaN  NaN  2.0   2  11300   0   0   2  
1  2.0  2.0  3.0   2   2208   0   0   2  
2  NaN  NaN  1.0   2      0   0   0   1  
3  3.0  5.3  2.0   1   2208   0   0   1  
4  NaN  NaN  2.0   2   4300   0   0   2  

[5 rows x 28 columns]

In [50]:
column_names = {0:'surgery', 1: 'age', 3: 'rectal_temperature', 4: 'pulse', 5: 'respiratory_rate', 6: 'temperature_of_extremities', 10: 'pain', 22: 'outcome'}

In [51]:
df = df[[0, 1, 3, 4, 5, 6, 10, 22]]

In [52]:
df_named = df.rename(columns = column_names)
df_named

surgery  age  rectal_temperature  pulse  respiratory_rate  \
0        2.0    1                38.5   66.0              28.0   
1        1.0    1                39.2   88.0              20.0   
2        2.0    1                38.3   40.0              24.0   
3        1.0    9                39.1  164.0              84.0   
4        2.0    1                37.3  104.0              35.0   
..       ...  ...                 ...    ...               ...   
295      1.0    1                 NaN  120.0              70.0   
296      2.0    1                37.2   72.0              24.0   
297      1.0    1                37.5   72.0              30.0   
298      1.0    1                36.5  100.0              24.0   
299      1.0    1                37.2   40.0              20.0   

     temperature_of_extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           NaN   NaN      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[300 rows x 8 columns]

In [53]:
df_named.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 8 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     299 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            242 non-null    float64
 5   temperature_of_extremities  244 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     299 non-null    float64
dtypes: float64(7), int64(1)
memory usage: 18.9 KB


In [54]:
df_named.loc[~df_named.age.isin([1,2])] 

surgery  age  rectal_temperature  pulse  respiratory_rate  \
3        1.0    9                39.1  164.0              84.0   
9        2.0    9                38.3   90.0               NaN   
13       2.0    9                38.0   92.0              28.0   
16       1.0    9                 NaN  128.0              36.0   
23       1.0    9                38.3  130.0              60.0   
39       1.0    9                39.2  146.0              96.0   
41       2.0    9                39.0  150.0              72.0   
55       1.0    9                38.6  160.0              20.0   
74       1.0    9                 NaN    NaN               NaN   
75       1.0    9                39.7  100.0               NaN   
82       1.0    9                38.1  100.0              80.0   
103      1.0    9                38.0  140.0              68.0   
109      1.0    9                38.4   84.0              40.0   
135      1.0    9                38.1  136.0              48.0   
191      2.0    9                 NaN  100.0              44.0   
212      1.0    9                38.9  120.0              30.0   
227      1.0    9                 NaN    NaN               NaN   
229      1.0    9                38.5  120.0              70.0   
244      1.0    9                38.2  124.0              88.0   
252      1.0    9                38.8  124.0              36.0   
255      1.0    9                38.8  184.0              84.0   
262      2.0    9                38.6   88.0              28.0   
275      1.0    9                38.8  150.0              50.0   
287      2.0    9                39.5   84.0              30.0   

     temperature_of_extremities  pain  outcome  
3                           4.0   2.0      2.0  
9                           1.0   5.0      1.0  
13                          1.0   1.0      2.0  
16                          3.0   4.0      2.0  
23                          NaN   2.0      1.0  
39                          NaN   NaN      2.0  
41                          NaN   NaN      1.0  
55                          3.0   3.0      2.0  
74                          NaN   NaN      2.0  
75                          3.0   2.0      3.0  
82                          3.0   3.0      1.0  
103                         1.0   3.0      1.0  
109                         3.0   3.0      2.0  
135                         3.0   5.0      2.0  
191                         2.0   4.0      1.0  
212                         1.0   3.0      1.0  
227                         NaN   NaN      1.0  
229                         NaN   NaN      1.0  
244                         1.0   2.0      1.0  
252                         3.0   2.0      2.0  
255                         1.0   4.0      2.0  
262                         NaN   NaN      1.0  
275                         1.0   5.0      2.0  
287                         NaN   NaN      1.0

Возраст не из указанного диапазона значений, таких строк - 24 из 300. Удалять из-за этого строки не хочется. Видим, что значений "2" нет в принципе, логично предположить, что 9 использовалось вместо 2.

In [55]:
df_named.loc[df_named.age == 2]

Empty DataFrame
Columns: [surgery, age, rectal_temperature, pulse, respiratory_rate, temperature_of_extremities, pain, outcome]
Index: []

In [56]:
df_named.loc[~df_named.age.isin([1,2]), 'age'] = 2

In [57]:
df_orig = df_named.copy()

Выбросы в столбцах respiratory_rate, rectal_temperature, pulse кажутся естественными, поэтому заменять или удалять их не будем, просто будем считать меры центральной тенденции без них.

In [58]:
def remove_outliers(column,df_named):
    '''
    удаляет выбросы в указанном столбце
    '''
    q1 = df_named[column].quantile(0.25)
    q3 = df_named[column].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr)
    upper_bound = q3 + (1.5 * iqr)
    df_named = df_named[df_named[column].between(lower_bound, upper_bound, inclusive = True)]
    return df_named

In [59]:
df_without_outliers = remove_outliers('respiratory_rate', df_named)
df_without_outliers.describe()

surgery         age  rectal_temperature       pulse  \
count  224.000000  225.000000          195.000000  219.000000   
mean     1.424107    1.053333           38.118462   68.716895   
std      0.495314    0.225198            0.726287   26.245178   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.750000   48.000000   
50%      1.000000    1.000000           38.100000   60.000000   
75%      2.000000    1.000000           38.500000   86.000000   
max      2.000000    2.000000           40.800000  160.000000   

       respiratory_rate  temperature_of_extremities        pain     outcome  
count        225.000000                  188.000000  191.000000  224.000000  
mean          26.715556                    2.351064    2.890052    1.540179  
std           11.461464                    1.051685    1.286961    0.738555  
min            8.000000                    1.000000    1.000000    1.000000  
25%           18.000000                    1.000000    2.000000    1.000000  
50%           24.000000                    3.000000    3.000000    1.000000  
75%           35.000000                    3.000000    4.000000    2.000000  
max           60.000000                    4.000000    5.000000    3.000000

In [60]:
df_named.describe()

surgery         age  rectal_temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.080000           38.167917   71.913043   
std      0.490305    0.271746            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

In [61]:
# respiratory_rate вообще считается сомнительным показателем, поэтому без зазрений совести и группировок заполним пропуски средним без учета выбросов
df_named['respiratory_rate'] = df_named['respiratory_rate'].fillna(df_without_outliers['respiratory_rate'].mean())

Пропуск в графе операция заполняем на 1 (была проведена), т.к. "- In general, the more painful, the more likely it is to require surgery" из описания, уровень боли высокий (4 из 5). Остальные показатели не сильно критичные, так что будем считать, что она выжила.

In [62]:
df_named['surgery'] = df_named['surgery'].fillna(1)

In [63]:
df_named['outcome'] = df_named['outcome'].fillna(1)

Hot extremities should correlate with an elevated rectal temp. normal temp is 37.8
Высокой ректальной темп будем считать от 38. Для высокой темп конечностей заполним ректальную температуру средним по повышенной темп, для повышенной темп заполним температуру конечностей 2 = Warm.
Нормальной температурой будем считать от 37 до 38, для нее температуру конечностей ставим 1 = Normal


In [64]:
def temp_classification(temp):
    '''классифицирует температуру на высокую, нормальную, низкую и добавляет соотв. столбец'''
    t_class = np.NaN
    if temp >= 38:
        t_class = 'high'
    if temp < 37:
        t_class = 'low'
    if temp>= 37 and temp <38:
        t_class = 'normal'
    return t_class    
        

In [65]:
df_named['temp_class'] = df_named['rectal_temperature'].apply(temp_classification)
df_named

surgery  age  rectal_temperature  pulse  respiratory_rate  \
0        2.0    1                38.5   66.0              28.0   
1        1.0    1                39.2   88.0              20.0   
2        2.0    1                38.3   40.0              24.0   
3        1.0    2                39.1  164.0              84.0   
4        2.0    1                37.3  104.0              35.0   
..       ...  ...                 ...    ...               ...   
295      1.0    1                 NaN  120.0              70.0   
296      2.0    1                37.2   72.0              24.0   
297      1.0    1                37.5   72.0              30.0   
298      1.0    1                36.5  100.0              24.0   
299      1.0    1                37.2   40.0              20.0   

     temperature_of_extremities  pain  outcome temp_class  
0                           3.0   5.0      2.0       high  
1                           NaN   3.0      3.0       high  
2                           1.0   3.0      1.0       high  
3                           4.0   2.0      2.0       high  
4                           NaN   NaN      2.0     normal  
..                          ...   ...      ...        ...  
295                         4.0   2.0      3.0        NaN  
296                         3.0   4.0      3.0     normal  
297                         4.0   4.0      2.0     normal  
298                         3.0   3.0      1.0        low  
299                         NaN   NaN      3.0     normal  

[300 rows x 9 columns]

In [66]:
df_without_outliers = remove_outliers('rectal_temperature', df_named)
df_without_outliers.describe()

surgery         age  rectal_temperature       pulse  \
count  226.00000  226.000000          226.000000  220.000000   
mean     1.40708    1.084071           38.169027   69.040909   
std      0.49238    0.278110            0.571949   28.673778   
min      1.00000    1.000000           36.800000   30.000000   
25%      1.00000    1.000000           37.800000   48.000000   
50%      1.00000    1.000000           38.200000   60.000000   
75%      2.00000    1.000000           38.500000   86.000000   
max      2.00000    2.000000           39.500000  184.000000   

       respiratory_rate  temperature_of_extremities        pain     outcome  
count        226.000000                  185.000000  187.000000  226.000000  
mean          29.758053                    2.210811    2.759358    1.473451  
std           17.321539                    1.070297    1.236043    0.719076  
min            8.000000                    1.000000    1.000000    1.000000  
25%           20.000000                    1.000000    2.000000    1.000000  
50%           26.715556                    2.000000    3.000000    1.000000  
75%           34.750000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000

In [67]:
df_without_outliers.groupby(['temp_class'])['temperature_of_extremities'].agg(pd.Series.mode)

temp_class
high      3.0
low       2.0
normal    1.0
Name: temperature_of_extremities, dtype: float64

Видим, что самые частые значения не соответствуют нашему предположению - при высокой температуре конечности чаще прохладные, а при низкой - теплые. Будем заполнять в итоге модой (которой почему-то нет в методе groupby)

In [68]:
def fill_temp_ex(row):
        if row['temp_class'] == 'high':
            return 3.0
        if row['temp_class'] == 'low':
            return 2.0   
        if row['temp_class'] == 'normal':
            return 1.0  
        else:
            return np.NaN

In [69]:
df_named['temperature_of_extremities'] = df_named.fillna(df_named.apply(fill_temp_ex, axis = 1))

In [70]:
df_named.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     300 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            300 non-null    float64
 5   temperature_of_extremities  300 non-null    object 
 6   pain                        245 non-null    float64
 7   outcome                     300 non-null    float64
 8   temp_class                  240 non-null    object 
dtypes: float64(6), int64(1), object(2)
memory usage: 21.2+ KB


In [71]:
df_named['temperature_of_extremities'] = pd.to_numeric(df_named['temperature_of_extremities'])

In [72]:
df_named

surgery  age  rectal_temperature  pulse  respiratory_rate  \
0        2.0    1                38.5   66.0              28.0   
1        1.0    1                39.2   88.0              20.0   
2        2.0    1                38.3   40.0              24.0   
3        1.0    2                39.1  164.0              84.0   
4        2.0    1                37.3  104.0              35.0   
..       ...  ...                 ...    ...               ...   
295      1.0    1                 NaN  120.0              70.0   
296      2.0    1                37.2   72.0              24.0   
297      1.0    1                37.5   72.0              30.0   
298      1.0    1                36.5  100.0              24.0   
299      1.0    1                37.2   40.0              20.0   

     temperature_of_extremities  pain  outcome temp_class  
0                           2.0   5.0      2.0       high  
1                           1.0   3.0      3.0       high  
2                           2.0   3.0      1.0       high  
3                           1.0   2.0      2.0       high  
4                           2.0   NaN      2.0     normal  
..                          ...   ...      ...        ...  
295                         1.0   2.0      3.0        NaN  
296                         2.0   4.0      3.0     normal  
297                         1.0   4.0      2.0     normal  
298                         1.0   3.0      1.0        low  
299                         1.0   NaN      3.0     normal  

[300 rows x 9 columns]

In [73]:
df_named.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300 entries, 0 to 299
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   surgery                     300 non-null    float64
 1   age                         300 non-null    int64  
 2   rectal_temperature          240 non-null    float64
 3   pulse                       276 non-null    float64
 4   respiratory_rate            300 non-null    float64
 5   temperature_of_extremities  300 non-null    float64
 6   pain                        245 non-null    float64
 7   outcome                     300 non-null    float64
 8   temp_class                  240 non-null    object 
dtypes: float64(7), int64(1), object(1)
memory usage: 21.2+ KB


Разделим пульс на категории: low - <30, normal - 30-40, medhigh - 40-90, high - >90

In [74]:
def pulse_classification(pulse):
    '''делит пульс на категории: low - <30, normal - 30-40, medhigh - 40-90, high - >90'''
    p_class = np.NaN
    if pulse >=90:
        p_class = 'high'
    if pulse < 30:
        p_class = 'low'
    if pulse >= 30 and pulse <=40:
        p_class = 'normal'
    if pulse > 40 and pulse <90:
        p_class = 'medhigh'    
    return p_class    

In [75]:
df_named['pulse_class'] = df_named['pulse'].apply(pulse_classification)
df_named

surgery  age  rectal_temperature  pulse  respiratory_rate  \
0        2.0    1                38.5   66.0              28.0   
1        1.0    1                39.2   88.0              20.0   
2        2.0    1                38.3   40.0              24.0   
3        1.0    2                39.1  164.0              84.0   
4        2.0    1                37.3  104.0              35.0   
..       ...  ...                 ...    ...               ...   
295      1.0    1                 NaN  120.0              70.0   
296      2.0    1                37.2   72.0              24.0   
297      1.0    1                37.5   72.0              30.0   
298      1.0    1                36.5  100.0              24.0   
299      1.0    1                37.2   40.0              20.0   

     temperature_of_extremities  pain  outcome temp_class pulse_class  
0                           2.0   5.0      2.0       high     medhigh  
1                           1.0   3.0      3.0       high     medhigh  
2                           2.0   3.0      1.0       high      normal  
3                           1.0   2.0      2.0       high        high  
4                           2.0   NaN      2.0     normal        high  
..                          ...   ...      ...        ...         ...  
295                         1.0   2.0      3.0        NaN        high  
296                         2.0   4.0      3.0     normal     medhigh  
297                         1.0   4.0      2.0     normal     medhigh  
298                         1.0   3.0      1.0        low        high  
299                         1.0   NaN      3.0     normal      normal  

[300 rows x 10 columns]

In [76]:
df_without_outliers = remove_outliers('pulse', df_named)

При повышенной температуре часто наблюдается повышение пульса, проверим эту теорию:

In [77]:
df_without_outliers.groupby(['pulse_class'])['rectal_temperature'].median()

pulse_class
high       38.4
medhigh    38.1
normal     38.0
Name: rectal_temperature, dtype: float64

Заполним недостающую темп мединным значением при группировке по пульсу, оставшиеся - просто медианными значениями 

In [78]:
df_named['rectal_temperature'] = df_named['rectal_temperature'].fillna(df_without_outliers.groupby(['pulse_class'])['rectal_temperature'].transform('median'))
df_named.describe()

surgery         age  rectal_temperature       pulse  \
count  300.000000  300.000000          282.000000  276.000000   
mean     1.396667    1.080000           38.173759   71.913043   
std      0.490023    0.271746            0.678091   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.100000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        300.000000                  300.000000  245.00000  300.000000  
mean          29.701674                    1.396667    2.95102    1.550000  
std           15.906497                    0.490023    1.30794    0.736642  
min            8.000000                    1.000000    1.00000    1.000000  
25%           20.000000                    1.000000    2.00000    1.000000  
50%           26.715556                    1.000000    3.00000    1.000000  
75%           34.250000                    2.000000    4.00000    2.000000  
max           96.000000                    2.000000    5.00000    3.000000

In [79]:
df_named['rectal_temperature'] = df_named['rectal_temperature'].fillna(df_without_outliers['rectal_temperature'].median())
df_named.describe()

surgery         age  rectal_temperature       pulse  \
count  300.000000  300.000000          300.000000  276.000000   
mean     1.396667    1.080000           38.169333   71.913043   
std      0.490023    0.271746            0.657597   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.100000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        300.000000                  300.000000  245.00000  300.000000  
mean          29.701674                    1.396667    2.95102    1.550000  
std           15.906497                    0.490023    1.30794    0.736642  
min            8.000000                    1.000000    1.00000    1.000000  
25%           20.000000                    1.000000    2.00000    1.000000  
50%           26.715556                    1.000000    3.00000    1.000000  
75%           34.250000                    2.000000    4.00000    2.000000  
max           96.000000                    2.000000    5.00000    3.000000

Заполним оставшийся пульс в зависимости от температуры. Мы видим, что лошадей с пониженной температурой было всего 9, и разброс пульса был большой. Все данные с низкой температурой попадают в выбросы. Поэтому будем заполнять нормальную и высокую темп средним пульсом по группе, а низкую - просто средним пульсом. 
                        

In [86]:
df_named.loc[df_named.temp_class == 'low']

surgery  age  rectal_temperature  pulse  respiratory_rate  \
44       1.0    1                35.4  140.0         24.000000   
80       1.0    1                36.4   98.0         35.000000   
113      1.0    1                36.8   60.0         28.000000   
118      1.0    1                36.5   78.0         30.000000   
141      2.0    1                36.0   42.0         30.000000   
238      2.0    1                36.1   88.0         26.715556   
251      2.0    1                36.6   42.0         18.000000   
277      2.0    1                36.9   50.0         40.000000   
298      1.0    1                36.5  100.0         24.000000   

     temperature_of_extremities  pain  outcome temp_class pulse_class  
44                          1.0   4.0      3.0        low        high  
80                          1.0   4.0      2.0        low        high  
113                         1.0   NaN      2.0        low     medhigh  
118                         1.0   5.0      1.0        low     medhigh  
141                         2.0   NaN      2.0        low     medhigh  
238                         2.0   3.0      3.0        low     medhigh  
251                         2.0   1.0      2.0        low     medhigh  
277                         2.0   1.0      1.0        low     medhigh  
298                         1.0   3.0      1.0        low        high

In [88]:
df_named.groupby(['temp_class'])['pulse'].mean().round(0)

temp_class
high      74.0
low       78.0
normal    62.0
Name: pulse, dtype: float64

In [80]:
df_without_outliers = remove_outliers('rectal_temperature', df_named)

In [87]:
df_without_outliers.groupby(['temp_class'])['pulse'].mean().round(0)

temp_class
high      73.0
normal    62.0
Name: pulse, dtype: float64

In [89]:
df_named['pulse'] = df_named['pulse'].fillna(df_without_outliers.groupby(['temp_class'])['pulse'].transform('mean').round(0))
df_named.describe()

surgery         age  rectal_temperature       pulse  \
count  300.000000  300.000000          300.000000  281.000000   
mean     1.396667    1.080000           38.169333   71.932384   
std      0.490023    0.271746            0.657597   28.374140   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.100000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        300.000000                  300.000000  245.00000  300.000000  
mean          29.701674                    1.396667    2.95102    1.550000  
std           15.906497                    0.490023    1.30794    0.736642  
min            8.000000                    1.000000    1.00000    1.000000  
25%           20.000000                    1.000000    2.00000    1.000000  
50%           26.715556                    1.000000    3.00000    1.000000  
75%           34.250000                    2.000000    4.00000    2.000000  
max           96.000000                    2.000000    5.00000    3.000000

In [92]:
df_named['pulse'] = df_named['pulse'].fillna(round(df_without_outliers['pulse'].mean(), 0))
df_named.describe()

surgery         age  rectal_temperature       pulse  \
count  300.000000  300.000000          300.000000  300.000000   
mean     1.396667    1.080000           38.169333   71.873333   
std      0.490023    0.271746            0.657597   27.458767   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.100000   68.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        300.000000                  300.000000  245.00000  300.000000  
mean          29.701674                    1.396667    2.95102    1.550000  
std           15.906497                    0.490023    1.30794    0.736642  
min            8.000000                    1.000000    1.00000    1.000000  
25%           20.000000                    1.000000    2.00000    1.000000  
50%           26.715556                    1.000000    3.00000    1.000000  
75%           34.250000                    2.000000    4.00000    2.000000  
max           96.000000                    2.000000    5.00000    3.000000

Осталось разобраться с уровнем боли. In general, the more painful, the more likely it is to require surgery. Попробуем группировку по тому, была ли проведена операция.

In [93]:
df_named.groupby(['surgery'])['pain'].agg(pd.Series.mode)

surgery
1.0    3.0
2.0    1.0
Name: pain, dtype: float64

In [95]:
df_named['pain'] = df_named['pain'].fillna(df_named.groupby(['surgery'])['pain'].transform(lambda x: x.mode()[0]))
df_named.describe()

surgery         age  rectal_temperature       pulse  \
count  300.000000  300.000000          300.000000  300.000000   
mean     1.396667    1.080000           38.169333   71.873333   
std      0.490023    0.271746            0.657597   27.458767   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.900000   48.000000   
50%      1.000000    1.000000           38.100000   68.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities        pain     outcome  
count        300.000000                  300.000000  300.000000  300.000000  
mean          29.701674                    1.396667    2.840000    1.550000  
std           15.906497                    0.490023    1.270083    0.736642  
min            8.000000                    1.000000    1.000000    1.000000  
25%           20.000000                    1.000000    2.000000    1.000000  
50%           26.715556                    1.000000    3.000000    1.000000  
75%           34.250000                    2.000000    4.000000    2.000000  
max           96.000000                    2.000000    5.000000    3.000000

In [96]:
df_orig.describe()

surgery         age  rectal_temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.080000           38.167917   71.913043   
std      0.490305    0.271746            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   

       respiratory_rate  temperature_of_extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000

Вывод: с температурой конечностей, видимо, была на самая правильная стратегия выбрана 